### ISSUE TITLE
Improve parallelization of large REST test suites

### ISSUE THREAD

**Issue by mark-vieira** [elastic/elasticsearch#78043]
This is mostly motivated by https://github.com/elastic/elasticsearch/issues/69973, which is that we have large YAML test suites that are increasingly growing as we add functionality. This monoliths cause all sorts of issues because a) we don't run anything in parallel and b) their execution times can exceed 30 minutes or more in some configurations. Our BWC jobs on `7.x` are now taking over an hour mainly due to the `:rest-api-spec` tests, and similarly `:qa:mixed-cluster` (which uses the same set of YAML tests).

Simply put, we need a solution for partitioning these test suites, and ideally, running them in parallel. The simplest solution would be to generate a bunch of separate test tasks, each of which executes some subset of the test suite. This is nasty in many ways but it also doesn't solve the problem since Gradle won't concurrently execute tasks within a single subproject. Generating subprojects isn't scalable either.

Another option is to parallelize at the test executor level. Essentially, leveraging `maxParallelForks` to run multiple "sub-suites" in parallel against the same test cluster. We'd have to ensure this is ok, and that parallel executing tests on the same cluster won't clobber eachother.

**Comment by mark-vieira** [comment]
@breskeby any thoughts on this? We need to get ahead of this before these test suites becoming completely unmanagable.

**Comment by mark-vieira** [comment]
FWIW, we can take this on incrementally, and focus on these specific projects that run the giant YAML test suite. Because of that we can potentially get more creative in ways that we don't have to solve every edge case for ever other project that runs rest tests. This could include leveraging fancy features in Junit 5, etc.

**Comment by mark-vieira** [comment]
That said, if `@ParallelizableTask` still existed it would make our lives so much easier. Do we have any idea what is holding up intra-project parallism for `Test` tasks? Not having this _severely_ limits our ability to parallelize BWC testing in particular.

**Comment by mark-vieira** [comment]
I might play around with leveraging JUnit 5 for the YAML testing stuff. There are somethings that would actually be really useful, such as using [dynamic tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dynamic-tests). Also intra-class parallelism is supported: https://junit.org/junit5/docs/current/user-guide/#writing-tests-parallel-execution

**Comment by breskeby** [comment]
I need to get myself a bit more familiar how this yaml tests work under the hood in order to have a good answer on this. Yeah @ParallizableTask got removed in favour of the Gradle Worker Api which doesn't help us here.

**Comment by breskeby** [comment]
It seems testing what junit5 can bring us here is the best angle to attach parallelism here IMO.

I remember to ultimately enable parallel task execution per project in Gradle `configuration-on-demand` was said to be the precondition for that resulting in immutable project configuration first. I'll reach out to the gradle folks at see what the status is here or if there is any way to bypass this. 
Parallel test execution per task in Gradle unfortunately doesn't work if only one test class is used as we do often in our YamlRestTests having one prominent test class like `MixedClusterClientYamlTestSuiteIT`. I wonder if junit5s dynamic tests work better here with gradle parallel test execution. Ideally it does as it it should be possible to implement support in the gradle junit5runner. If not I guess junits intra class parallelism support is the best way to go. 
We could reach out to Marc Phillips to see what's possible here. 

Nevertheless, It would still be interesting to know and nail down why those 6.8 tests are soo slow.

**Comment by mark-vieira** [comment]
>  I wonder if junit5s dynamic tests work better here with gradle parallel test execution.

Yeah, I'm not sure it will a I think all dynamic tests will be considered part of the same class. I'm not aware of a good way to generate test _classes_ beyond actual code generation.

> If not I guess junits intra class parallelism support is the best way to go.

Yeah, that was my thought as well. I _think_ we can just enable this out of the box :shrug:

**Comment by pugnascotia** [comment]
@mark-vieira is this still something we want to do?

**Comment by mark-vieira** [comment]
Yes, but it's a pretty big effort.
